File: packages/python/src/softschema/generate.py
"""Generated schema sections: deterministic Markdown blocks from `SchemaView`.

Marker format:

    <!-- softschema:generated kind="enum_table" schema="schemas/foo.schema.yaml" -->
    ...rendered content...
    <!-- /softschema:generated -->

The renderer parses the attributes from the opening marker, loads the referenced
compiled schema through `SchemaView`, renders the section, and writes the file back.
With ``check=True`` it returns a drift report instead of touching the file, so CI
can fail when committed sections lag behind the schema. The path attribute is
``schema=`` (a file path); the older ``contract=`` spelling is rejected, since a
contract is a logical ID, not a file path.

Kinds shipped in v0.1:

- ``enum_table``: one table row per enum-valued field in the schema.
- ``field_list``: one bullet per field (name, type, required, description).
- ``vocab``: the enum values at a specific JSON Pointer (requires ``pointer=``).
"""

from __future__ import annotations

import re
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path

from strif import atomic_write_text

from softschema.schema_view import SchemaView

_MARKER_OPEN = re.compile(
    r"<!--\s*softschema:generated\s+(?P<attrs>[^>]*?)\s*-->",
)
_MARKER_CLOSE = "<!-- /softschema:generated -->"
_ATTR_PATTERN = re.compile(r'(?P<key>[A-Za-z_][A-Za-z0-9_]*)="(?P<value>[^"]*)"')
_KNOWN_ATTRS = frozenset({"kind", "schema", "pointer", "sha256"})
_CONTRACT_ATTR_RENAME = (
    'softschema:generated marker uses the removed "contract" attribute for a schema '
    'path; rename it to "schema" (contract is a logical ID, not a file path)'
)


@dataclass
class GeneratedSection:
    """One parsed marker block from a document."""

    start: int  # offset just after the opening marker (start of content)
    end: int  # offset of the closing marker
    open_marker: str
    close_marker: str
    attrs: dict[str, str]
    existing_content: str

    @property
    def kind(self) -> str:
        return self.attrs.get("kind", "")

    @property
    def schema(self) -> str:
        return self.attrs.get("schema", "")


@dataclass
class RegenerateResult:
    """Outcome of regenerating one document."""

    path: Path
    sections: int = 0
    drift: bool = False
    drift_details: list[str] = field(default_factory=list)


def parse_sections(text: str) -> list[GeneratedSection]:
    """Find every ``softschema:generated`` marker block in ``text``."""
    sections: list[GeneratedSection] = []
    cursor = 0
    while True:
        open_match = _MARKER_OPEN.search(text, cursor)
        if open_match is None:
            return sections
        open_end = open_match.end()
        close_index = text.find(_MARKER_CLOSE, open_end)
        if close_index == -1:
            msg = (
                f"unterminated softschema:generated marker starting at offset {open_match.start()}"
            )
            raise ValueError(msg)
        attrs = _parse_attrs(open_match.group("attrs"))
        existing = text[open_end:close_index]
        sections.append(
            GeneratedSection(
                start=open_end,
                end=close_index,
                open_marker=open_match.group(0),
                close_marker=_MARKER_CLOSE,
                attrs=attrs,
                existing_content=existing,
            )
        )
        cursor = close_index + len(_MARKER_CLOSE)


def regenerate(
    path: Path,
    *,
    check: bool = False,
    schema_root: Path | None = None,
) -> RegenerateResult:
    """Regenerate every ``softschema:generated`` block in ``path``.

    With ``check=True`` the file is not touched; the result reports whether any
    block has drifted from what the schema would currently render.

    ``schema_root`` controls how the ``schema="..."`` attribute is resolved: by
    default a relative path is resolved relative to the file containing the marker.
    """
    text = path.read_text(encoding="utf-8")
    sections = parse_sections(text)
    result = RegenerateResult(path=path, sections=len(sections))
    if not sections:
        return result

    base = schema_root or path.parent
    new_text = ""
    cursor = 0
    for section in sections:
        rendered = _render_section(section, base)
        new_text += text[cursor : section.start] + rendered
        cursor = section.end
        if rendered != section.existing_content:
            result.drift = True
            result.drift_details.append(f"{path}: section {section.attrs.get('kind', '?')} drifted")
    new_text += text[cursor:]

    if not check and result.drift:
        atomic_write_text(path, new_text)
    return result


def _render_section(section: GeneratedSection, schema_root: Path) -> str:
    if "contract" in section.attrs:
        raise ValueError(_CONTRACT_ATTR_RENAME)
    unknown = sorted(key for key in section.attrs if key not in _KNOWN_ATTRS)
    if unknown:
        msg = f"softschema:generated marker has unknown attribute(s): {', '.join(unknown)}"
        raise ValueError(msg)
    if not section.kind:
        msg = f"softschema:generated marker is missing 'kind': {section.open_marker}"
        raise ValueError(msg)
    if not section.schema:
        msg = f"softschema:generated marker is missing 'schema': {section.open_marker}"
        raise ValueError(msg)
    renderer = _RENDERERS.get(section.kind)
    if renderer is None:
        known = ", ".join(sorted(_RENDERERS))
        msg = f"unknown softschema:generated kind {section.kind!r}; known: {known}"
        raise ValueError(msg)
    schema_path = _resolve_schema_path(section.schema, schema_root)
    view = SchemaView.load(schema_path)
    body = renderer(view, section.attrs)
    return f"\n{body}\n"


def _resolve_schema_path(schema: str, base: Path) -> Path:
    path = Path(schema)
    if not path.is_absolute():
        path = base / path
    if not path.exists():
        msg = f"softschema:generated schema not found: {path}"
        raise FileNotFoundError(msg)
    return path


def _render_enum_table(view: SchemaView, attrs: dict[str, str]) -> str:
    lines = ["| Field | Allowed values |", "| --- | --- |"]
    for field_info in view.iter_fields():
        if field_info.enum is None:
            continue
        values = ", ".join(v.replace("|", "\\|") for v in field_info.enum)
        lines.append(f"| `{field_info.name}` | {values} |")
    if len(lines) == 2:
        lines.append("| _(no enum fields)_ | _(none)_ |")
    return "\n".join(lines)


def _render_field_list(view: SchemaView, attrs: dict[str, str]) -> str:
    lines: list[str] = []
    for field_info in view.iter_fields():
        # Only list root-level properties for readability; nested properties show
        # up via their parent's type.
        if field_info.pointer.count("/properties/") > 1:
            continue
        type_label = field_info.json_type or "object"
        required = "required" if field_info.required else "optional"
        description = field_info.description or ""
        if description:
            lines.append(f"- `{field_info.name}` ({type_label}, {required}): {description}")
        else:
            lines.append(f"- `{field_info.name}` ({type_label}, {required})")
    if not lines:
        lines.append("- _(no fields)_")
    return "\n".join(lines)


def _render_vocab(view: SchemaView, attrs: dict[str, str]) -> str:
    pointer = attrs.get("pointer")
    if not pointer:
        msg = "softschema:generated kind=vocab requires a 'pointer' attribute"
        raise ValueError(msg)
    values = view.enum_values(pointer)
    if values is None:
        msg = f"no enum at pointer {pointer!r}"
        raise ValueError(msg)
    return "\n".join(f"- `{v}`" for v in values)


_RENDERERS: dict[str, Callable[[SchemaView, dict[str, str]], str]] = {
    "enum_table": _render_enum_table,
    "field_list": _render_field_list,
    "vocab": _render_vocab,
}


def _parse_attrs(attr_string: str) -> dict[str, str]:
    return {
        match.group("key"): match.group("value") for match in _ATTR_PATTERN.finditer(attr_string)
    }

File: packages/typescript/src/generate.ts
/**
 * Generated schema sections: deterministic Markdown blocks rendered from a SchemaView.
 * Idiomatic mirror of the Python `generate.py`; renderer output is byte-identical so a
 * `softschema generate` on either CLI produces the same marker bodies.
 */
import { readFileSync } from "node:fs";
import { dirname, isAbsolute, join } from "node:path";
import { writeFileSync } from "atomically";
import { SchemaView } from "./schemaView.js";

const MARKER_OPEN = /<!--\s*softschema:generated\s+([^>]*?)\s*-->/g;
const MARKER_CLOSE = "<!-- /softschema:generated -->";
const ATTR_PATTERN = /([A-Za-z_][A-Za-z0-9_]*)="([^"]*)"/g;
const KNOWN_ATTRS = new Set(["kind", "schema", "pointer", "sha256"]);
const CONTRACT_ATTR_RENAME =
  'softschema:generated marker uses the removed "contract" attribute for a schema ' +
  'path; rename it to "schema" (contract is a logical ID, not a file path)';

export interface GeneratedSection {
  start: number;
  end: number;
  attrs: Record<string, string>;
  existingContent: string;
}

export interface RegenerateResult {
  path: string;
  sections: number;
  drift: boolean;
  driftDetails: string[];
}

function parseAttrs(attrString: string): Record<string, string> {
  const attrs: Record<string, string> = {};
  for (const match of attrString.matchAll(ATTR_PATTERN)) {
    attrs[match[1] as string] = match[2] as string;
  }
  return attrs;
}

export function parseSections(text: string): GeneratedSection[] {
  const sections: GeneratedSection[] = [];
  MARKER_OPEN.lastIndex = 0;
  let openMatch: RegExpExecArray | null = MARKER_OPEN.exec(text);
  while (openMatch !== null) {
    const contentStart = openMatch.index + openMatch[0].length;
    const closeIndex = text.indexOf(MARKER_CLOSE, contentStart);
    if (closeIndex === -1) {
      throw new Error(
        `unterminated softschema:generated marker starting at offset ${openMatch.index}`,
      );
    }
    sections.push({
      start: contentStart,
      end: closeIndex,
      attrs: parseAttrs(openMatch[1] as string),
      existingContent: text.slice(contentStart, closeIndex),
    });
    MARKER_OPEN.lastIndex = closeIndex + MARKER_CLOSE.length;
    openMatch = MARKER_OPEN.exec(text);
  }
  return sections;
}

function resolveSchemaPath(schema: string, base: string): string {
  const path = isAbsolute(schema) ? schema : join(base, schema);
  return path;
}

function renderEnumTable(view: SchemaView): string {
  const lines = ["| Field | Allowed values |", "| --- | --- |"];
  for (const fieldInfo of view.iterFields()) {
    if (fieldInfo.enum === null) continue;
    lines.push(`| \`${fieldInfo.name}\` | ${fieldInfo.enum.join(", ")} |`);
  }
  if (lines.length === 2) lines.push("| _(no enum fields)_ | _(none)_ |");
  return lines.join("\n");
}

function renderFieldList(view: SchemaView): string {
  const lines: string[] = [];
  for (const fieldInfo of view.iterFields()) {
    if (fieldInfo.pointer.split("/properties/").length - 1 > 1) continue;
    const typeLabel = fieldInfo.jsonType ?? "object";
    const required = fieldInfo.required ? "required" : "optional";
    const description = fieldInfo.description ?? "";
    if (description) {
      lines.push(`- \`${fieldInfo.name}\` (${typeLabel}, ${required}): ${description}`);
    } else {
      lines.push(`- \`${fieldInfo.name}\` (${typeLabel}, ${required})`);
    }
  }
  if (lines.length === 0) lines.push("- _(no fields)_");
  return lines.join("\n");
}

function renderVocab(view: SchemaView, attrs: Record<string, string>): string {
  const pointer = attrs.pointer;
  if (!pointer) throw new Error("softschema:generated kind=vocab requires a 'pointer' attribute");
  const values = view.enumValues(pointer);
  if (values === null) throw new Error(`no enum at pointer ${JSON.stringify(pointer)}`);
  return values.map((v) => `- \`${v}\``).join("\n");
}

const RENDERERS: Record<string, (view: SchemaView, attrs: Record<string, string>) => string> = {
  enum_table: (view) => renderEnumTable(view),
  field_list: (view) => renderFieldList(view),
  vocab: renderVocab,
};

function renderSection(section: GeneratedSection, schemaRoot: string): string {
  if ("contract" in section.attrs) throw new Error(CONTRACT_ATTR_RENAME);
  const unknown = Object.keys(section.attrs)
    .filter((key) => !KNOWN_ATTRS.has(key))
    .sort();
  if (unknown.length > 0) {
    throw new Error(`softschema:generated marker has unknown attribute(s): ${unknown.join(", ")}`);
  }
  const kind = section.attrs.kind ?? "";
  const schema = section.attrs.schema ?? "";
  if (!kind) throw new Error("softschema:generated marker is missing 'kind'");
  if (!schema) throw new Error("softschema:generated marker is missing 'schema'");
  const renderer = RENDERERS[kind];
  if (renderer === undefined) {
    const known = Object.keys(RENDERERS).sort().join(", ");
    throw new Error(`unknown softschema:generated kind ${JSON.stringify(kind)}; known: ${known}`);
  }
  const schemaPath = resolveSchemaPath(schema, schemaRoot);
  const view = SchemaView.load(schemaPath);
  return `\n${renderer(view, section.attrs)}\n`;
}

export function regenerate(
  path: string,
  options: { check?: boolean; schemaRoot?: string } = {},
): RegenerateResult {
  const text = readFileSync(path, "utf8");
  const sections = parseSections(text);
  const result: RegenerateResult = {
    path,
    sections: sections.length,
    drift: false,
    driftDetails: [],
  };
  if (sections.length === 0) return result;

  const base = options.schemaRoot ?? dirname(path);
  let newText = "";
  let cursor = 0;
  for (const section of sections) {
    const rendered = renderSection(section, base);
    newText += text.slice(cursor, section.start) + rendered;
    cursor = section.end;
    if (rendered !== section.existingContent) {
      result.drift = true;
      result.driftDetails.push(`${path}: section ${section.attrs.kind ?? "?"} drifted`);
    }
  }
  newText += text.slice(cursor);

  if (!options.check && result.drift) {
    writeFileSync(path, newText, { encoding: "utf8" });
  }
  return result;
}
